frontend/model/contracts/group.js

```diff
@@ -313,38 +325,43 @@ sbp('chelonia/defineContract', {
       return getters.groupSettings.mincomeCurrency
     },
     periodStampGivenDate (state, getters) {
-      return (recentDate: string | Date) => {
-        if (typeof recentDate !== 'string') {
-          recentDate = recentDate.toISOString()
+      return (date: Date | string): string | null => {
+        const givenStamp = typeof date === 'string' ? date : date.toISOString()
+        // Should we throw instead?
+        if (!isPeriodStamp(givenStamp)) return null
+        let currentPeriodStamp = getters.groupSettings.distributionDate
+        // Not sure whether this can happen. Keeping it anyway.
+        if (!currentPeriodStamp) return null
+        let nextPeriodStamp = getters.periodAfterPeriod(currentPeriodStamp)
+        let previousPeriodStamp = getters.periodBeforePeriod(currentPeriodStamp)
+        if (compareISOTimestamps(givenStamp, currentPeriodStamp) >= 0) {
+          // Find the first period stamp that's later than the given date.
+          while (nextPeriodStamp && compareISOTimestamps(givenStamp, nextPeriodStamp) >= 0) {
+            currentPeriodStamp = nextPeriodStamp
+            nextPeriodStamp = getters.periodAfterPeriod(nextPeriodStamp)
+          }
+          return currentPeriodStamp
+        } else {
+          // Iterate backwards to find a first previous period stamp that's lower than the given date.
+          while (previousPeriodStamp && compareISOTimestamps(givenStamp, previousPeriodStamp) < 0) {
+            previousPeriodStamp = getters.periodBeforePeriod(previousPeriodStamp)
+          }
+          // Now either 'previousPeriodStamp' is undefined or 'givenString' >= 'previousPeriodStamp'.
+          return previousPeriodStamp ?? null
         }
-        const { distributionDate, distributionPeriodLength } = getters.groupSettings
-
-        if (!distributionDate) return null
-
-        return periodStampGivenDate({
-          recentDate,
-          periodStart: distributionDate,
-          periodLength: distributionPeriodLength
-        })
       }
     },
-    periodBeforePeriod (state, getters) {
-      return (periodStamp: string) => {
-        const len = getters.groupSettings.distributionPeriodLength
-        return dateToPeriodStamp(addTimeToDate(dateFromPeriodStamp(periodStamp), -len))
-      }
+    periodBeforePeriod (state, getters): string | void {
+      return (periodStamp: string) => getters.groupPeriodPayments[periodStamp]?.previousPeriodID
     },
-    periodAfterPeriod (state, getters) {
-      return (periodStamp: string) => {
-        const len = getters.groupSettings.distributionPeriodLength
-        return dateToPeriodStamp(addTimeToDate(dateFromPeriodStamp(periodStamp), len))
-      }
+    periodAfterPeriod (state, getters): string | void {
+      return (periodStamp: string) => getters.groupPeriodPayments[periodStamp]?.nextPeriodID
```

Comment:
This needs to be modified because sometimes we are given dates in the future, for example, have a look at this code in `state.js`:

```js
const reactiveDate = Vue.observable({ date: new Date() })
setInterval(function () {
  // We want the getters to recalculate all of the payments within 1 minute of us entering a new period.
  const rememberedPeriodStamp = store.getters.periodStampGivenDate?.(reactiveDate.date)
  const currentPeriodStamp = store.getters.periodStampGivenDate?.(new Date())
  if (rememberedPeriodStamp !== currentPeriodStamp) {
    reactiveDate.date = new Date()
  }
}, 60 * 1000)
```

This code will break with the way `periodAfterPeriod` is currently implemented, but it would work if you re-introduced the code we have before. You might want to combine the two, e.g. if `nextPeriodID` doesn't exist, use the old code to generate the correct periodStamp using the `distributionPeriodLength`